src/connectors/clear-refinements/connectClearRefinements.js

```diff
@@ -89,70 +91,62 @@ export default function connectClearRefinements(renderFn, unmountFn = noop) {
     return {
       $$type: 'ais.clearRefinements',
 
-      init({ helper, instantSearchInstance, createURL }) {
-        const attributesToClear = getAttributesToClear({
-          helper,
-          includedAttributes,
-          excludedAttributes,
-          transformItems,
-        });
-        const hasRefinements = attributesToClear.length > 0;
-
-        this._refine = () => {
-          helper
-            .setState(
-              clearRefinements({
-                helper,
-                attributesToClear: getAttributesToClear({
-                  helper,
-                  includedAttributes,
-                  excludedAttributes,
-                  transformItems,
-                }),
-              })
-            )
-            .search();
-        };
-
-        this._createURL = () =>
-          createURL(
-            clearRefinements({
-              helper,
-              attributesToClear: getAttributesToClear({
-                helper,
-                includedAttributes,
-                excludedAttributes,
-                transformItems,
-              }),
-            })
-          );
-
+      init({ instantSearchInstance }) {
         renderFn(
           {
-            hasRefinements,
-            refine: this._refine,
-            createURL: this._createURL,
+            hasRefinements: false,
+            refine: noop,
+            createURL: noop,
             instantSearchInstance,
             widgetParams,
           },
           true
         );
       },
 
-      render({ helper, instantSearchInstance }) {
-        const attributesToClear = getAttributesToClear({
-          helper,
-          includedAttributes,
-          excludedAttributes,
-          transformItems,
-        });
-        const hasRefinements = attributesToClear.length > 0;
+      render({ scopedResults, createURL, instantSearchInstance }) {
+        const attributesToClear = scopedResults.reduce(
+          (results, scopedResult) => {
+            return results.concat(
+              getAttributesToClear({
+                scopedResult,
+                includedAttributes,
+                excludedAttributes,
+                transformItems,
+              })
+            );
+          },
+          []
+        );
 
         renderFn(
           {
-            hasRefinements,
-            refine: this._refine,
-            createURL: this._createURL,
+            hasRefinements: attributesToClear.some(
+              attributeToClear => attributeToClear.items.length > 0
+            ),
+            refine: () => {
```

Comment:
With those changes we provide a new function on each render which was not the case before. It can lead to issues when we the render function have a branch for the first render. This problem arises because the `scopedResults` is not provided at the `init` step. A solution might be to provide it, even empty it doesn't matter.

With an empty state for `scopedResults` we can create the function at the `init` step like before. It will have the same behavior than the proposed implementation the call to `refine`/`createURL` will be a noop.

Reply:
That's right, I fixed that in 07a010772b1c3fc3c6c5e3a715dd703bb0fa5e4f. I went for storing this in the connector state and assigning noops like this implementation.

What do you mean by this though?

> It can lead to issues when we the render function have a branch for the first render.

On a side note, it would be nice to have the same signature for both `init` and `render` at some point.

Reply:
> What do you mean by this though?

The ref change between render which means that if we have a render function that branch on `isFirstRender` to register events, the function created at the render step will never be called.

Reply:
The fix made inside https://github.com/algolia/instantsearch.js/commit/07a010772b1c3fc3c6c5e3a715dd703bb0fa5e4f does not solve the problem. The reference is different since the function gets re-assigned. You can take a look at [`refinementList`](https://github.com/algolia/instantsearch.js/blob/a23c301f1909d67516664108ff8e015cdea21516/src/connectors/refinement-list/connectRefinementList.js#L285-L290) or [`geoSearch`](https://github.com/algolia/instantsearch.js/blob/a23c301f1909d67516664108ff8e015cdea21516/src/connectors/geo-search/connectGeoSearch.js#L220-L226), both connectors use different strategies to keep the reference.

Reply:
I get it now. This is addressed in 72a39a6fb7f293c512ef3974d492e8a4ec514470.